### DISCUSSION TITLE
Form autocompletion for a field with lots of options

### DISCUSSION THREAD

**Discussion by Murat60** [symfony/symfony#42798] [Q&A] [answered]
Bonjour je suis sur un projet symfony 5 et je voudrais réalisé sur mon formulaire gérer par une vue twig un champs pour sélectionner la ville de l'utilisateur depuis ma base de donnée.

Quel serait le moyen le plus facile pour avoir un champs dynamique pour évidemment ne pas charger la liste des centaines de ville sur ma page.

Merci beaucoup

**Comment by ruudk** [comment]
Please use English.

**Reply by Murat60** [reply]
Hello ruudk i am on a symfony 5 project and I would like to realize on my form with twig view a field to select the user's city from my database.

What would be the easiest way to have a dynamic field so obviously not to load the list of hundreds of cities on my page.

thank you so much

**Reply by ruudk** [reply]
You have to do some JavaScript.

Maybe check the notes below the video for inspiration: https://symfonycasts.com/screencast/symfony-forms/autocomplete-ajax

**Comment by jfsenechal** [comment, marked as answer]
You can use stimulus 

https://symfonycasts.com/screencast/stimulus
https://github.com/afcapel/stimulus-autocomplete

Example from scratch : 

**Install encore** 

```bash
composer require encore
```
**Add module autocomplete**

```bash
yarn add stimulus-autocomplete
```

**Active module**

```js
//assets/bootstrap.js
import { startStimulusApp } from '@symfony/stimulus-bridge';
import { Autocomplete } from 'stimulus-autocomplete';

//.....

app.register('autocomplete', Autocomplete);
```

**Compile js**

```bash
yarn install
yarn build
```
**Add js to your pages**

```twig
//templates/base.html.twig
//in head section
        {% block javascripts %}
            {{ encore_entry_script_tags('app') }}
        {% endblock %}
    </head>
```

**Create query ajax**

```php
    /**
     * @Route("/ajax/town/", name="app_ajax_town", methods={"GET"})
     */
    public function ajaxTown(Request $request): Response
    {
        $keyword = $request->get('q');
        $towns = [];
        if ($keyword) {
            $towns = $this->townRepository->findByName($keyword);
        }
        
        return $this->render('_list.html.twig', ['towns' => $towns]);
    }
```
**Template twig**

```twig
{% for town in towns %}
    <li class="list-group-item" role="option" data-autocomplete-value="{{ town.id }}">
            {{ town.name }}
    </li>
{% endfor %}
```

**Form**

```php
class SearchNameType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('townid', HiddenType::class, [
                'attr' => ['data-autocomplete-target' => 'hidden'],
            ])
            ->add('name', SearchType::class, [
                'attr' => ['data-autocomplete-target' => 'input'],
            ]);
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            // Configure your form options here
        ]);
    }
}
```
**Render your form and call autocomplete**

```twig
    <div class="example-wrapper" {{ stimulus_controller('autocomplete', {'url': path('app_ajax_town')}) }}>
        {{ form(form) }}
        <ul class="list-group" data-autocomplete-target="results"></ul>
    </div>
```
If you want get only name and not id, remove data-autocomplete-value="{{ town.id }}" from template twig
